// @nuxt/ui/dist/module.mjs
import { addTemplate, useNuxt, installModule, defineNuxtModule, createResolver, addPlugin, addComponentsDir, addImportsDir } from '@nuxt/kit';
import { createRequire } from 'node:module';
import { join } from 'pathe';
import { kebabCase } from 'scule';



// -- Unbuild CommonJS Shims --
import __cjs_url__ from 'url';
import __cjs_path__ from 'path';
import __cjs_mod__ from 'module';
const __filename = __cjs_url__.fileURLToPath(import.meta.url);
const __dirname = __cjs_path__.dirname(__filename);
const require = __cjs_mod__.createRequire(import.meta.url);
const name = "@nuxt/ui";
const version = "2.14.0";

function createTemplates(nuxt = useNuxt()) {
  const template = addTemplate({
    filename: "ui.colors.mjs",
    getContents: () => `export default ${JSON.stringify(nuxt.options.appConfig.ui.colors)};`,
    write: true
  });
  const typesTemplate = addTemplate({
    filename: "ui.colors.d.ts",
    getContents: () => `declare module '#ui-colors' { const defaultExport: ${JSON.stringify(nuxt.options.appConfig.ui.colors)}; export default defaultExport; }`,
    write: true
  });
  nuxt.options.alias["#ui-colors"] = template.dst;
  nuxt.hook("prepare:types", (opts) => {
    opts.references.push({ path: typesTemplate.dst });
  });
}

function omit(object, keysToOmit) {
  const result = { ...object };
  for (const key of keysToOmit) {
    delete result[key];
  }
  return result;
}

const colorsToExclude = [
  "inherit",
  "transparent",
  "current",
  "white",
  "black",
  "slate",
  "gray",
  "zinc",
  "neutral",
  "stone",
  "cool"
];
const safelistByComponent = {
  alert: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-50`)
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`)
  }],
  avatar: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }],
  badge: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-50`)
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`)
  }],
  button: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-50`),
    variants: ["hover", "disabled"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-100`),
    variants: ["hover"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark", "dark:disabled"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`),
    variants: ["disabled", "dark:hover"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-600`),
    variants: ["hover"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-900`),
    variants: ["dark:hover"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-950`),
    variants: ["dark", "dark:hover", "dark:disabled"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark", "dark:disabled"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`),
    variants: ["dark:hover", "disabled"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-600`),
    variants: ["hover"]
  }, {
    pattern: new RegExp(`outline-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`outline-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }],
  input: (colorsAsRegex2) => [{
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark", "dark:focus"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus"]
  }],
  radio: (colorsAsRegex2) => [{
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }],
  checkbox: (colorsAsRegex2) => [{
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }],
  toggle: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }],
  range: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-400`),
    variants: ["dark:focus-visible"]
  }, {
    pattern: new RegExp(`ring-(${colorsAsRegex2})-500`),
    variants: ["focus-visible"]
  }],
  progress: (colorsAsRegex2) => [{
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }],
  meter: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }],
  notification: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`text-(${colorsAsRegex2})-500`)
  }],
  chip: (colorsAsRegex2) => [{
    pattern: new RegExp(`bg-(${colorsAsRegex2})-400`),
    variants: ["dark"]
  }, {
    pattern: new RegExp(`bg-(${colorsAsRegex2})-500`)
  }]
};
const colorsAsRegex = (colors) => colors.join("|");
const excludeColors = (colors) => {
  return Object.entries(omit(colors, colorsToExclude)).filter(([, value]) => typeof value === "object").map(([key]) => kebabCase(key));
};
const generateSafelist = (colors, globalColors) => {
  const baseSafelist = Object.keys(safelistByComponent).flatMap((component) => safelistByComponent[component](colorsAsRegex(colors)));
  const formsSafelist = ["input", "radio", "checkbox", "toggle", "range"].flatMap((component) => safelistByComponent[component](colorsAsRegex(["red"])));
  return [
    ...baseSafelist,
    ...formsSafelist,
    // Ensure all global colors are safelisted for the Notification (toast.add)
    ...safelistByComponent["notification"](colorsAsRegex(globalColors)),
    // Gray safelist for Avatar & Notification
    "bg-gray-500",
    "dark:bg-gray-400",
    "text-gray-500",
    "dark:text-gray-400"
  ];
};

const _require = createRequire(import.meta.url);
const defaultColors = _require("tailwindcss/colors.js");
delete defaultColors.lightBlue;
delete defaultColors.warmGray;
delete defaultColors.trueGray;
delete defaultColors.coolGray;
delete defaultColors.blueGray;
async function installTailwind(moduleOptions, nuxt = useNuxt(), { resolve, runtimeDir }) {
  nuxt.hook("tailwindcss:config", function(tailwindConfig) {
    tailwindConfig.theme = tailwindConfig.theme || {};
    tailwindConfig.theme.extend = tailwindConfig.theme.extend || {};
    tailwindConfig.theme.extend.colors = tailwindConfig.theme.extend.colors || {};
    const globalColors = {
      ...tailwindConfig.theme.colors || defaultColors,
      ...tailwindConfig.theme.extend?.colors
    };
    globalColors.primary = tailwindConfig.theme.extend.colors.primary = {
      50: "rgb(var(--color-primary-50) / <alpha-value>)",
      100: "rgb(var(--color-primary-100) / <alpha-value>)",
      200: "rgb(var(--color-primary-200) / <alpha-value>)",
      300: "rgb(var(--color-primary-300) / <alpha-value>)",
      400: "rgb(var(--color-primary-400) / <alpha-value>)",
      500: "rgb(var(--color-primary-500) / <alpha-value>)",
      600: "rgb(var(--color-primary-600) / <alpha-value>)",
      700: "rgb(var(--color-primary-700) / <alpha-value>)",
      800: "rgb(var(--color-primary-800) / <alpha-value>)",
      900: "rgb(var(--color-primary-900) / <alpha-value>)",
      950: "rgb(var(--color-primary-950) / <alpha-value>)",
      DEFAULT: "rgb(var(--color-primary-DEFAULT) / <alpha-value>)"
    };
    if (globalColors.gray) {
      globalColors.cool = tailwindConfig.theme.extend.colors.cool = defaultColors.gray;
    }
    globalColors.gray = tailwindConfig.theme.extend.colors.gray = {
      50: "rgb(var(--color-gray-50) / <alpha-value>)",
      100: "rgb(var(--color-gray-100) / <alpha-value>)",
      200: "rgb(var(--color-gray-200) / <alpha-value>)",
      300: "rgb(var(--color-gray-300) / <alpha-value>)",
      400: "rgb(var(--color-gray-400) / <alpha-value>)",
      500: "rgb(var(--color-gray-500) / <alpha-value>)",
      600: "rgb(var(--color-gray-600) / <alpha-value>)",
      700: "rgb(var(--color-gray-700) / <alpha-value>)",
      800: "rgb(var(--color-gray-800) / <alpha-value>)",
      900: "rgb(var(--color-gray-900) / <alpha-value>)",
      950: "rgb(var(--color-gray-950) / <alpha-value>)"
    };
    const colors = excludeColors(globalColors);
    nuxt.options.appConfig.ui = {
      primary: "green",
      gray: "cool",
      colors,
      strategy: "merge"
    };
    tailwindConfig.safelist = tailwindConfig.safelist || [];
    tailwindConfig.safelist.push(
      ...generateSafelist(moduleOptions.safelistColors || [], colors)
    );
  });
  const configTemplate = addTemplate({
    filename: "nuxtui-tailwind.config.cjs",
    write: true,
    getContents: () => `
      const { defaultExtractor: createDefaultExtractor } = require('tailwindcss/lib/lib/defaultExtractor.js')
      const { customSafelistExtractor } = require(${JSON.stringify(resolve(runtimeDir, "utils", "colors"))})
      const { iconsPlugin, getIconCollections } = require('@egoist/tailwindcss-icons')

      const defaultExtractor = createDefaultExtractor({ tailwindConfig: { separator: ':' } })

      module.exports = {
        darkMode: 'class',
        plugins: [
          require('@tailwindcss/forms')({ strategy: 'class' }),
          require('@tailwindcss/aspect-ratio'),
          require('@tailwindcss/typography'),
          require('@tailwindcss/container-queries'),
          require('@headlessui/tailwindcss'),
          iconsPlugin(${Array.isArray(moduleOptions.icons) || moduleOptions.icons === "all" ? `{ collections: getIconCollections(${JSON.stringify(moduleOptions.icons)}) }` : typeof moduleOptions.icons === "object" ? JSON.stringify(moduleOptions.icons) : "{}"})
        ],
        content: {
          files: [
            ${JSON.stringify(resolve(runtimeDir, "components/**/*.{vue,mjs,ts}"))},
            ${JSON.stringify(resolve(runtimeDir, "ui.config/**/*.{mjs,js,ts}"))}
          ],
          transform: {
            vue: (content) => {
              return content.replaceAll(/(?:\\r\\n|\\r|\\n)/g, ' ')
            }
          },
          extract: {
            vue: (content) => {
              return [
                ...defaultExtractor(content),
                ...customSafelistExtractor(${JSON.stringify(moduleOptions.prefix)}, content, ${JSON.stringify(nuxt.options.appConfig.ui.colors)}, ${JSON.stringify(moduleOptions.safelistColors)})
              ]
            }
          }
        }
      }
    `
  });
  await installModule("@nuxtjs/tailwindcss", {
    exposeConfig: true,
    configPath: [
      configTemplate.dst,
      join(nuxt.options.rootDir, "tailwind.config")
    ]
  });
}

const module = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "ui",
    compatibility: {
      nuxt: "^3.10.0"
    }
  },
  defaults: {
    prefix: "U",
    icons: ["heroicons"],
    safelistColors: ["primary"],
    disableGlobalStyles: false
  },
  async setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    const runtimeDir = resolve("./runtime");
    nuxt.options.build.transpile.push(runtimeDir);
    nuxt.options.build.transpile.push("@popperjs/core", "@headlessui/vue");
    nuxt.options.alias["#ui"] = runtimeDir;
    if (!options.disableGlobalStyles) {
      nuxt.options.css.push(resolve(runtimeDir, "ui.css"));
    }
    createTemplates(nuxt);
    await installModule("nuxt-icon");
    await installModule("@nuxtjs/color-mode", { classSuffix: "" });
    await installTailwind(options, nuxt, { resolve, runtimeDir });
    addPlugin({
      src: resolve(runtimeDir, "plugins", "colors")
    });
    addPlugin({
      src: resolve(runtimeDir, "plugins", "modals")
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "elements"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "forms"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "data"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "layout"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "navigation"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addComponentsDir({
      path: resolve(runtimeDir, "components", "overlays"),
      prefix: options.prefix,
      global: options.global,
      watch: false
    });
    addImportsDir(resolve(runtimeDir, "composables"));
  }
});

export { module as default };
